Add flask-admin and flask-security

sonalranjit opened the pull request:
Goals:

- Set up admin user accounts (no registration option for general users at this point). Role types should include:
  - Admins (can add/remove users and edit data)
  - Maintainers (can edit data and manage their own account)
- Set up and test admin panels for logging in and managing accounts

Future PRs:

- Set up versioning in the `bikeparking_submissions` table to allow for updates with recorded history
- Add a status table linked to `bikeparking_submissions`

Todos:

- [ ] security review
- [x] update tests and check coverage
- [x] set-up flow: add secrets for seed user

sonalranjit committed:
Working base version of flask-admin and flask-security

sonalranjit committed:
Separate modelviews according to models

sonalranjit committed:
Disable self registration

tallcoleman committed:
Developer experience - enable and document use of .env for flask-security secrets

sonalranjit committed:
Generate a migration file for the new user models

tallcoleman committed:
point api docker compose files to .env

sonalranjit committed:
Set password field to be type of password to hide content

sonalranjit committed:
Update unit test to account for new redirection of landling page

tallcoleman committed:
Add manage script to create seed superuser

adds script to manage.py to create superuser if there are no admins in the database using secrets set via .env or deployment environment

tallcoleman committed:
Increased memory allocation in fly.toml

app had out of memory error with 256mb, but 512mb seems OK

tallcoleman committed:
Fix env variable configuration for github actions

tallcoleman committed:
set env_file as optional in docker compose

tallcoleman commented:
@sonalranjit added a script to add a seed superuser using environment secrets and updated the tests so that they are now passing

sonalranjit committed:
Started adding tests for admin panel

sonalranjit committed:
Add more functional test

sonalranjit committed:
Add more functional tests for admin page

tallcoleman committed:
Add pytest-dotenv

Allows .env file to be used to set secrets when tests are run locally

tallcoleman commented on bikespace_api/bikespace_api/api/forms.py:
Is this being used or can this file be deleted?

tallcoleman committed:
Add tests for string representations

Bring coverage for models.py back to 100%

tallcoleman commented on bikespace_api/bikespace_api/api/models.py:
```diff
@@ -6,7 +6,8 @@
 from enum import Enum
-
+from flask_security import RoleMixin, UserMixin
+from werkzeug.security import generate_password_hash, check_password_hash
```
In my editor, it shows the following errors for the imports:

- sqlalchemy.sql.func - import not used
- RoleMixin and UserMixin should be imported from flask_security.core
- werkzeug.security imports not used

sonalranjit replied:
hmm strange, i'll look at it

tallcoleman committed:
Ignore pywright error for current_user import

See: https://github.com/pallets-eco/flask-security/discussions/1089

tallcoleman commented on bikespace_api/bikespace_api/api/views.py:
```diff
@@ -0,0 +1,107 @@
+from flask_security import current_user
+from flask_security.utils import hash_password
+from wtforms import SelectField, PasswordField
+from bikespace_api.api.models import IssueType, ParkingDuration
```
the bikespace_api.api.models imports as well as the wtforms.SelectField import appear to be un-used

sonalranjit replied:
Yupp, seems to be unused, i'll remove it

tallcoleman committed:
Fix class inheriting from self

tallcoleman commented on bikespace_api/bikespace_api/api/views.py:
A lot of the class methods (e.g. `is_accessible` and `_handle_view`) are the same or very similar for these views. Would it make sense to make a custom subclass to use instead of ModelView?

See: https://github.com/pallets-eco/flask-admin/blob/a9f2e0c418f57f97b3f1a59bb8680473418a1c56/examples/auth/main.py#L88

tallcoleman commented on bikespace_api/bikespace_api/templates/security/register_user.html:
I'm assuming this template is not currently being used but can be for future updates?

sonalranjit replied:
Yeah not used, im not sure if we want a self registering page

tallcoleman committed:
ran make lint-py

tallcoleman committed:
fix typo

tallcoleman commented on bikespace_api/bikespace_api/api/views.py:
```diff
@@ -0,0 +1,107 @@
+from bikespace_api.api.models import IssueType, ParkingDuration
+import uuid
+
+class AdminIndexView(AdminIndexView):
```
When I run `make test-api-terminal` it shows that there's no coverage for this class even though `test_admin_page.py` is calling the admin index page both logged in and not. Do you know why this is the case?

tallcoleman replied:
see: https://github.com/bikespace/bikespace/pull/276#discussion_r2460192707

tallcoleman committed:
Updated pytest fixtures

Changed to more closely match the set-up recommended in the Flask documentation (providing separate app and client fixtures and managing context persistence on a per-test basis).

Also added some notes on using and configuring pytest fixtures in the README.

tallcoleman committed:
Test allowed pages for regular users

- need to disable CSRF during testing
- fixtures should not have module scope (otherwise admin login in one test bleeds over into the client for anothr test in the same module); static fixtures do not need scope also

tallcoleman committed:
Add type ignore for bs4 find function

tallcoleman committed:
Add test coverage for AdminUsersModelView

Covers AdminUsersModelView.on_model_change() to ensure that creating a user and updating their password using the flask-admin endpoints works correctly.

Also added a utility function that can be used while writing or debugging tests to pretty-print html.

tallcoleman commented on bikespace_api/bikespace_api/api/views.py:
```diff
@@ -0,0 +1,115 @@
+import uuid
+
+
+class MyAdminIndexView(AdminIndexView):
```
This function appears to not get any coverage (i.e. is not getting used) even if you call `test_client.get("/admin/")` with a superuser logged in. I think that it needs to be added when `Admin` is configured in the `__init__.py` `create_app` function.

I think if implemented, a regular user could still technically log in, but all the admin pages (including the root "/admin/" would 403 for them. The `test_allowed_pages_for_regular_users` function may need to be updated if that's the intended behaviour.

sonalranjit replied:
Yeah i removed this view, doesn't seem to do anything

tallcoleman commented on bikespace_api/bikespace_api/tests/conftest.py:
FYI I updated the way the fixtures are set up to align with the documentation in https://flask.palletsprojects.com/en/stable/testing/

- having module scope for the test client meant that if a user was logged in in one test, they would also be logged in for others, which led to some confusing consequences for tests in the same module that test login and non-login states, including for users with different permissions
- app context (e.g. required for tests that use direct db queries) can be provisioned as-needed using the flask_app fixture instead of being provided every time the test client is used

tallcoleman commented on bikespace_api/bikespace_api/config.py:
```diff
@@ -11,8 +11,28 @@ class BaseConfig:
+    # Flask-Security features
+    SECURITY_REGISTERABLE = False
+    SECURITY_SEND_REGISTER_EMAIL = False
+    SQLALCHEMY_TRACK_MODIFICATIONS = False
```
What does this flag do?

sonalranjit replied:
These disable self registering an admin account. We don't want that.

tallcoleman commented on bikespace_api/bikespace_api/config.py:
```diff
@@ -26,6 +46,7 @@ class TestingConfig(BaseConfig):
 
     TESTING = True
     SQLALCHEMY_DATABASE_URI = os.environ.get("TEST_DATABASE_URI")
+    WTF_CSRF_ENABLED = False
```
Added this to the testing config since there are some endpoints that would otherwise require a CSRF token - there are ways to have your tests also get and use a CSRF token, but that's additional complexity we could look at adding in later.

tallcoleman commented on bikespace_api/manage.py:
```diff
@@ -78,13 +104,69 @@ def seed_db():
     )
     db.session.commit()
 
+    # add roles for seeded users if they do not yet exist
```
Something worth looking into in the future - I think we can do some of this set-up using test fixtures instead; might be cleaner. For now, this is fine though!

tallcoleman reviewed (changes requested):
@sonalranjit thanks for all your hard work on this first step towards enabling us to clean up our data and better track the status of issues!

FYI I added some more test coverage to your changes, but it's still slightly below what it was before, so the coveralls action is unhappy. If we're not able to get it back over 98% again, you might have to bypass that check to merge or re-config the coveralls check?

I still want to make sure we do our clean-up non-destructively (i.e. tagging test submissions instead of outright deleting them from the db), so I'll plan to do a quick follow-up PR to make that possible.

tallcoleman committed:
Add name to migration

tallcoleman committed:
Show id in submissions admin panel view

sonalranjit commented:
<img width="898" height="372" alt="Screenshot 2025-10-28 at 5 50 14 PM" src="https://github.com/user-attachments/assets/adfbb3ed-8401-4b1d-a6f0-2ad5142edf8e" />
Something is up with Flask-Admin and WTForms version, seems like Flask-Admin `1.6.1` isn't compatible with WTForms `3.1.2` When i go to create a new Role it throws an error.

tallcoleman committed:
Bump API version

sonalranjit committed:
Bump flask-admin and WTForms versions

sonalranjit committed:
Update Flask-admin constructor with new version changes

sonalranjit committed:
Update role/user models to handle nullable string fields

sonalranjit committed:
Rename base html template for admin panel

sonalranjit committed:
Update functional test with the Flask-admin version changes

sonalranjit committed:
Remove unused import Modelview

sonalranjit committed:
Remove unused imports

sonalranjit committed:
Remove unused imports and unused AdminIndexView

sonalranjit commented:
This is ready to be reviewed again @tallcoleman, i've deployed it on a test deployment on fly, it seems to be working

tallcoleman committed:
Fix skipped evaluation in test

- if condition for test_admin_page was failing due to extra whitespace around string being searched; changed assert so that it should fail properly if not found
- changed all string= parameters to use regex match to avoid failing in cases where whitespace around value or capitalization changes
- removed unused imports

tallcoleman reviewed (approved):
@sonalranjit I think this is good to merge now. I pushed one fix to one of the tests since the test coverage was indicating that the assert condition wasn't getting called.

sonalranjit commented:
Awesome thanks @tallcoleman
